### ISSUE TITLE
Segfault in icmp send

### ISSUE THREAD

**Issue by 0xdeafbeef** [libpnet/libpnet#449]
https://github.com/0xd34b33f/cloudflare-ping/blob/06e359d16fff09839cc6a09b1ebdcacb9788ee91/src/ping_core.rs#L332
Segfault, when sending icmp packet. Works in debug mode or in --march=native. Doesn't work in release mode

**Comment by kishiguro** [comment]
Let me check.

**Comment by 0xd34b33f** [comment]
Tested under Linux and macOS. On Linux it works, on macOS it segfaults or doesn't work at all

**Comment by 0xd34b33f** [comment]
```
WRITE of size 40 at 0x7fff69dd9dc8 thread T0
    #0 0x55eb86108acf  (/home/username/oss/cloudflare-ping/target/x86_64-unknown-linux-gnu/release/internship_application_systems+0x1f6acf)
    #1 0x55eb861506a4  (/home/username/oss/cloudflare-ping/target/x86_64-unknown-linux-gnu/release/internship_application_systems+0x23e6a4)
    #2 0x55eb86153a87  (/home/username/oss/cloudflare-ping/target/x86_64-unknown-linux-gnu/release/internship_application_systems+0x241a87)
    #3 0x55eb8613c35d  (/home/username/oss/cloudflare-ping/target/x86_64-unknown-linux-gnu/release/internship_application_systems+0x22a35d)
    #4 0x55eb861322a2  (/home/username/oss/cloudflare-ping/target/x86_64-unknown-linux-gnu/release/internship_application_systems+0x2202a2)
    #5 0x55eb865dfbfc  (/home/username/oss/cloudflare-ping/target/x86_64-unknown-linux-gnu/release/internship_application_systems+0x6cdbfc)
    #6 0x55eb865e1688  (/home/username/oss/cloudflare-ping/target/x86_64-unknown-linux-gnu/release/internship_application_systems+0x6cf688)
    #7 0x55eb865df80f  (/home/username/oss/cloudflare-ping/target/x86_64-unknown-linux-gnu/release/internship_application_systems+0x6cd80f)
    #8 0x55eb865e13bc  (/home/username/oss/cloudflare-ping/target/x86_64-unknown-linux-gnu/release/internship_application_systems+0x6cf3bc)
    #9 0x55eb86132231  (/home/username/oss/cloudflare-ping/target/x86_64-unknown-linux-gnu/release/internship_application_systems+0x220231)
    #10 0x7fd879c26001  (/usr/lib/libc.so.6+0x27001)
    #11 0x55eb8609c2dd  (/home/username/oss/cloudflare-ping/target/x86_64-unknown-linux-gnu/release/internship_application_systems+0x18a2dd)

Address 0x7fff69dd9dc8 is located in stack of thread T0 at offset 3752 in frame
    #0 0x55eb8614e6af  (/home/username/oss/cloudflare-ping/target/x86_64-unknown-linux-gnu/release/internship_application_systems+0x23c6af)

  This frame has 80 object(s):
    [32, 40) 'item.i.i.i.i.i.i'
    [64, 72) 'bucket.i.i.i.i.i'
    [96, 120) '_5.i.i.i.i.i'
    [160, 184) '_4.i.i.i.i.i'
    [224, 256) '_7.i.i133' (line 235)
    [288, 320) 'replay.i.i' (line 235)
    [352, 376) 'vec.i.i'
    [416, 448) '_117.i' (line 292)
    [480, 528) '_110.i' (line 292)
    [560, 562) 'seq1.i' (line 291)
    [576, 600) '_69.i' (line 272)
    [640, 648) '_65.i' (line 272)
    [672, 752) '_61.i' (line 272)
    [784, 832) '_54.i' (line 272)
    [864, 944) 'formatter.i' (line 267)
    [976, 1000) '_39.i' (line 265)
    [1040, 1041) 'ttl.i' (line 255)
    [1056, 1058) 'seq.i' (line 255)
    [1072, 1080) 'size.i' (line 252)
    [1104, 1136) 'generic_icmp.i' (line 251)
    [1168, 1188) '_6.i126'
    [1232, 1264) '_5.i'
    [1296, 1328) '_13.i.i'
    [1360, 1392) '_8.i.i'
    [1424, 1425) 'val.i38.i'
    [1440, 1444) 'val.i19.i'
    [1456, 1457) 'val.i.i'
    [1472, 1504) 'icmp_packet.i' (line 53)
    [1536, 1560) 'e.i110'
    [1600, 1656) '_2.i.i.i'
    [1696, 1720) 'e.i'
    [1760, 1776) '_12.i.i.i'
    [1792, 1840) '_5.i.i.i'
    [1872, 1880) 'self.i.i.i'
    [1904, 1920) '_9.i'
    [1936, 1960) '_7.i'
    [2000, 2016) '_267' (line 384)
    [2032, 2080) '_260' (line 384)
    [2112, 2136) 'e8' (line 380)
    [2176, 2200) '_244' (line 380)
    [2240, 2264) '_242' (line 380)
    [2304, 2320) '_232' (line 378)
    [2336, 2384) '_225' (line 378)
    [2416, 2440) 'e6' (line 377)
    [2480, 2504) '_212' (line 377)
    [2544, 2568) '_210' (line 377)
    [2608, 2624) '_203' (line 375)
    [2640, 2688) '_196' (line 375)
    [2720, 2744) '_188' (line 368)
    [2784, 2804) '_168' (line 354)
    [2848, 2880) '_167' (line 354)
    [2912, 2968) 'a3' (line 352)
    [3008, 3072) '_156' (line 352)
    [3104, 3128) '_142' (line 349)
    [3168, 3216) '_119' (line 342)
    [3248, 3296) '_112' (line 342)
    [3328, 3344) 'e' (line 342)
    [3360, 3392) '_86' (line 341)
    [3424, 3472) '_79' (line 341)
    [3504, 3528) '_63' (line 340)
    [3568, 3600) 'packet' (line 330)
    [3632, 3672) 'buffer_icmp' (line 329)
    [3712, 3752) 'buffer_ip' (line 328) <== Memory access at offset 3752 overflows this variable
    [3792, 3848) '_48' (line 319) <== Memory access at offset 3752 underflows this variable
    [3888, 3912) '_47' (line 319)
    [3952, 3976) 'address_for_handler' (line 317)
    [4016, 4048) '_42' (line 316)
    [4080, 4096) 'ctrlc_rx' (line 315)
    [4112, 4128) 'ctrlc_tx' (line 315)
    [4144, 4176) '_28' (line 311)
    [4208, 4256) '_21' (line 311)
    [4288, 4292) 'addr' (line 307)
    [4304, 4306) 'seq_n' (line 306)
    [4320, 4376) 'ping_time' (line 305)
    [4416, 4424) 'rx1' (line 304)
    [4448, 4512) '_6' (line 300)
    [4544, 4600) '_5.sroa.5' (line 300)
    [4640, 4696) '_4' (line 300)
    [4736, 4776) 'rx' (line 300)
    [4816, 4832) 'tx' (line 300)
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-overflow (/home/username/oss/cloudflare-ping/target/x86_64-unknown-linux-gnu/release/internship_application_systems+0x1f6acf) 
Shadow bytes around the buggy address:
  0x10006d3b3360: f8 f8 f8 f8 f2 f2 f2 f2 f8 f8 f8 f2 f2 f2 f2 f2
  0x10006d3b3370: f8 f8 f8 f8 f8 f8 f2 f2 f2 f2 f8 f8 f8 f8 f8 f8
  0x10006d3b3380: f2 f2 f2 f2 f8 f8 f2 f2 f8 f8 f8 f8 f2 f2 f2 f2
  0x10006d3b3390: f8 f8 f8 f8 f8 f8 f2 f2 f2 f2 f8 f8 f8 f2 f2 f2
  0x10006d3b33a0: f2 f2 00 00 00 00 f2 f2 f2 f2 00 00 00 00 00 f2
=>0x10006d3b33b0: f2 f2 f2 f2 00 00 00 00 00[f2]f2 f2 f2 f2 f8 f8
  0x10006d3b33c0: f8 f8 f8 f8 f8 f2 f2 f2 f2 f2 f8 f8 f8 f2 f2 f2
  0x10006d3b33d0: f2 f2 00 00 00 f2 f2 f2 f2 f2 f8 f8 f8 f8 f2 f2
  0x10006d3b33e0: f2 f2 00 00 f2 f2 00 00 f2 f2 f8 f8 f8 f8 f2 f2
  0x10006d3b33f0: f2 f2 f8 f8 f8 f8 f8 f8 f2 f2 f2 f2 04 f2 02 f2
  0x10006d3b3400: 00 00 00 00 00 00 00 f2 f2 f2 f2 f2 00 f2 f2 f2
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==116168==ABORTING
```

**Comment by 0xd34b33f** [comment]
```
= Invalid read of size 4
==134127==    at 0x23BF53: pnet_transport::IcmpTransportChannelIterator::next_with_timeout (lib.rs:355)
==134127==    by 0x154C67: internship_application_systems::ping_core::Pinger::ping_ipv4 (ping_core.rs:352)
==134127==    by 0x155F4D: internship_application_systems::ping_core::Pinger::run (ping_core.rs:390)
==134127==    by 0x14C307: process (main.rs:141)
==134127==    by 0x14C307: internship_application_systems::main (main.rs:97)
==134127==    by 0x148C12: std::rt::lang_start::{{closure}} (rt.rs:67)
==134127==    by 0x2B56F7: {{closure}} (rt.rs:52)
==134127==    by 0x2B56F7: do_call<closure-0,i32> (panicking.rs:296)
==134127==    by 0x2B56F7: try<i32,closure-0> (panicking.rs:273)
==134127==    by 0x2B56F7: catch_unwind<closure-0,i32> (panic.rs:394)
==134127==    by 0x2B56F7: std::rt::lang_start_internal (rt.rs:51)
==134127==    by 0x14C757: main (in /home/username/oss/cloudflare-ping/target/release/internship_application_systems)
==134127==  Address 0x10 is not stack'd, malloc'd or (recently) free'd
==134127== 
==134127== 
==134127== Process terminating with default action of signal 11 (SIGSEGV): dumping core
==134127==  Access not within mapped region at address 0x10
==134127==    at 0x23BF53: pnet_transport::IcmpTransportChannelIterator::next_with_timeout (lib.rs:355)
==134127==    by 0x154C67: internship_application_systems::ping_core::Pinger::ping_ipv4 (ping_core.rs:352)
==134127==    by 0x155F4D: internship_application_systems::ping_core::Pinger::run (ping_core.rs:390)
==134127==    by 0x14C307: process (main.rs:141)
==134127==    by 0x14C307: internship_application_systems::main (main.rs:97)
==134127==    by 0x148C12: std::rt::lang_start::{{closure}} (rt.rs:67)
==134127==    by 0x2B56F7: {{closure}} (rt.rs:52)
==134127==    by 0x2B56F7: do_call<closure-0,i32> (panicking.rs:296)
==134127==    by 0x2B56F7: try<i32,closure-0> (panicking.rs:273)
==134127==    by 0x2B56F7: catch_unwind<closure-0,i32> (panic.rs:394)
==134127==    by 0x2B56F7: std::rt::lang_start_internal (rt.rs:51)
==134127==    by 0x14C757: main (in /home/username/oss/cloudflare-ping/target/release/internship_application_systems)
==134127==  If you believe this happened as a result of a stack
==134127==  overflow in your program's main thread (unlikely but
==134127==  possible), you can try to increase the size of the
==134127==  main thread stack using the --main-stacksize= flag.
==134127==  The main thread stack size used in this run was 8388608.
```

**Comment by kbknapp** [comment]
I took a look at this out of curiosity and it looks like this is because for some reason Rust is optimizing out the `rx` local variable in [`cloudflare-ping/src/ping_core.rs:304`](https://github.com/0xd34b33f/cloudflare-ping/blob/06e359d16fff09839cc6a09b1ebdcacb9788ee91/src/ping_core.rs#L304)

**Comment by kbknapp** [comment]
I'm on a Linux system (Fedora 32). I compiled `cloudflare-ping` in release mode with debug symbols included:

```toml
# cloudflare-ping Cargo.toml
[profile.release]
debug = true
```

Then compile and start in GDB, and running the following:

```
(gdb) break ping_core.rs:304
(gdb) run 1.1.1.1
(gdb) print rx
(gdb) next
(gdb) next
(gdb) print rx
```

Which results in this screenshot:

![2020-07-09_22-49](https://user-images.githubusercontent.com/6942134/87111235-d8d3c080-c236-11ea-90f5-b9b530154a66.png)

**Comment by 0xd34b33f** [comment]
What I  found out, it fails on that [line](https://github.com/libpnet/libpnet/blob/3e76647c42649c5ca6bb8bbc08074894d9457a54/pnet_transport/src/lib.rs#L355) trying to dereference fd of underlying socket. I tried to create a mre, but it works as expected and no segfaults :( @kbknapp, maybe it's problem with compiler itself?

**Comment by kbknapp** [comment]
> maybe it's problem with compiler itself?

Could be, but it could also be something libpnet is doing that hints to the compiler this is safe. For example there are many times throughout the `std` documentation where it's stated that certain uses of the API will invoke UB because the compiler makes assumptions for certain optimizations. See the [`std::mem::MaybeUninit`](https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#initialization-invariant) docs for example.

It might be worth filing an issue on `rustc` to see if they notice something that's being done to invoke this type of optimization.

**Comment by 0xdeafbeef** [comment]
@kishiguro what you think, should I create an issue on rustsec for it?

**Comment by kishiguro** [comment]
let's create an issue on rustsec for it.  we might be able to have help on the issue.  @0xdeafbeef would you mind to do the effort?

**Comment by 0xdeafbeef** [comment]
@kishiguro I'll do it

**Comment by 0xdeafbeef** [comment]
[Done](https://github.com/RustSec/advisory-db/pull/335)

**Comment by Shnatsel** [comment]
Thanks for the RustSec PR! Sadly the WG maintaining it cannot really help at this time, but here are other places to get help:

1. [Rust Community Discord](https://discord.gg/aVESxV8) has `#black-magic` channel with people who are knowledgeable about this sort of thing and are usually willing to help out
1. Rust [forums](https://users.rust-lang.org/) or [reddit](https://www.reddit.com/r/rust/) can get you a lot of eyeballs
1. Opening an issue against rustc is also an option if a cause does not present itself otherwise

**Comment by 0xdeafbeef** [comment]
Error in this method. https://github.com/libpnet/libpnet/blob/3e76647c42649c5ca6bb8bbc08074894d9457a54/pnet_transport/src/lib.rs#L355. 
```
                println!("{:#?}",&self.tr);
```
When I invoke this, `println!()` segfaults while formatting

**Comment by tdryer** [comment]
This looks like a buffer overrun. The application [uses 40 byte buffers](https://github.com/0xdeafbeef/cloudflare-ping/blob/06e359d16fff09839cc6a09b1ebdcacb9788ee91/src/ping_core.rs#L328-L329) for both the IPv4 and ICMP data. When [`ipv4_packet.set_payload` is called](https://github.com/0xdeafbeef/cloudflare-ping/blob/06e359d16fff09839cc6a09b1ebdcacb9788ee91/src/ping_core.rs#L59
), it copies the ICMP data into the IPv4 buffer, after the IPv4 header. Since both buffers are the same size, and the length of the IPv4 header is greater than zero, data is copied past the end of the IPv4 buffer. 

The [code in libpnet performing the copy](https://github.com/libpnet/libpnet/blob/3e76647c42649c5ca6bb8bbc08074894d9457a54/pnet_macros/src/decorator.rs#L663-L667) is:
```
unsafe {{
    copy_nonoverlapping(vals[..].as_ptr(),
                        _self.packet[current_offset..].as_mut_ptr(),
                        vals.len())
}}
```

Setting a breakpoint there seems to confirm that the copy will overflow `buffer_ip`. Both pointers are to 40 byte buffers, and `current_offset` is 20 bytes:
```
(gdb) break ipv4.rs:1002
(gdb) r
Thread 1 "internship_appl" hit Breakpoint 1, pnet_packet::ipv4::MutableIpv4Packet::set_payload (
    self=0x7fffffff8cc8, vals=...)
    at /home/tom/Programming/rust/cloudflare-ping/target/debug/build/pnet_packet-2c357403620fe001/out/ipv4.rs:1002
1002                copy_nonoverlapping(vals[..].as_ptr(),
(gdb) print vals
$1 = &[u8] {data_ptr: 0x7fffffff9330 "\b\000", length: 40}
(gdb) print self.packet
$2 = pnet_macros_support::packet::MutPacketData::Borrowed(&mut [u8] {data_ptr: 0x7fffffff9300 "E\000", length: 40})
(gdb) print current_offset
$3 = 20
(gdb)
```

There's another open issue #411 indicating that `set_payload` is unsound.

**Comment by FlyingRatBull** [comment]
As per [the Rust security advisory DB](https://github.com/RustSec/advisory-db) this bug should be fixed in 0.26.0 but this issue is still open and the pull request not yet merged. Is this still an issue in 0.26.0 or should I create a pull request to update the advisory DB?

**Comment by mrmonday** [comment]
It is not fixed in `0.26.0`, I'll rebase the PR and if CI passes release a `0.27.0` with the fix.

The advisory DB should probably be against the `pnet_transport` crate rather than `pnet`.

**Comment by FlyingRatBull** [comment]
Thank you for your quick help!
